diff -ru yum-3.2.22-orig/yum/config.py yum-3.2.22/yum/config.py
--- yum-3.2.22-orig/yum/config.py	2009-12-17 09:57:48.000000000 -0500
+++ yum-3.2.22/yum/config.py	2009-12-17 10:00:29.000000000 -0500
@@ -668,7 +668,7 @@
                  # all == install any/all arches you can
                  # best == use the 'best  arch' for the system
                  
-    bugtracker_url = Option('http://yum.baseurl.org/report')
+    bugtracker_url = Option('https://bugzilla.redhat.com/enter_bug.cgi?product=Red%20Hat%20Enterprise%20Linux%206&component=yum')
 
     color = SelectionOption('auto', ('auto', 'never', 'always'),
                             mapper={'on' : 'always', 'yes' : 'always',
Only in yum-3.2.22/yum: config.py~
