commit 4c8b223552a1bd3b830cef9dbfaf46012279dacd
Author: James Antill <james@and.org>
Date:   Mon Apr 11 13:22:58 2011 -0400

    Do basename checking of ssl cert. files. BZ 695427.

diff --git a/yum/__init__.py b/yum/__init__.py
index cf4d827..2b1db8e 100644
--- a/yum/__init__.py
+++ b/yum/__init__.py
@@ -597,6 +597,24 @@ class YumBase(depsolve.Depsolve):
 
 
         if doSetup:
+            if (hasattr(urlgrabber, 'grabber') and
+                hasattr(urlgrabber.grabber, 'pycurl')):
+                # Must do basename checking, on cert. files...
+                cert_basenames = {}
+                for repo in self._repos.listEnabled():
+                    if not repo.sslclientcert:
+                        continue
+                    bn = os.path.basename(repo.sslclientcert)
+                    if bn not in cert_basenames:
+                        cert_basenames[bn] = repo
+                        continue
+                    if repo.sslclientcert == cert_basenames[bn].sslclientcert:
+                        # Exactly the same path is fine too
+                        continue
+
+                    msg = 'sslclientcert basename shared between %s and %s'
+                    raise Errors.ConfigError, msg % (repo, cert_basenames[bn])
+
             repo_st = time.time()        
             self._repos.doSetup(thisrepo)
             self.verbose_logger.debug('repo time: %0.3f' % (time.time() - repo_st))        
commit 01ac40a22c491b7f2de37b5a34f3fc294b39d5b2
Author: James Antill <james@and.org>
Date:   Mon Apr 11 13:31:55 2011 -0400

    Document basename checking of ssl cert. files. BZ 695427.

diff --git a/docs/yum.conf.5 b/docs/yum.conf.5
index a535b79..2e0d6ea 100644
--- a/docs/yum.conf.5
+++ b/docs/yum.conf.5
@@ -318,6 +318,13 @@ is not registered.
 Path to the SSL client certificate yum should use to connect to repos/remote sites
 Defaults to none.
 
+Note that if you are using curl compiled against NSS (default in Fedora/RHEL),
+curl treats sslclientcert values with the same basename as _identical_. This
+version of yum will check that this isn't true and output an error when the
+repositories "foo" and "bar" violate this, like so:
+
+sslclientcert basename shared between foo and bar
+
 .IP
 \fBsslclientkey \fR
 Path to the SSL client key yum should use to connect to repos/remote sites
