commit 2a5d185b921f028b0216edec42beb41efef97061
Author: James Antill <james@and.org>
Date:   Fri Feb 4 11:24:41 2011 -0500

    Hide invalid XML errors for updateinfo, BZ 663378

diff --git a/yum/update_md.py b/yum/update_md.py
index 83e56c6..619250e 100644
--- a/yum/update_md.py
+++ b/yum/update_md.py
@@ -32,6 +32,15 @@ import Errors
 
 import rpmUtils.miscutils
 
+
+def safe_iterparse(filename):
+    """ Works like iterparse, but hides XML errors (prints a warning). """
+    try:
+        for event, elem in safe_iterparse(infile):
+            yield event, elem
+    except SyntaxError: # Bad XML
+        print >> sys.stderr, "File is not valid XML:", filename
+
 class UpdateNoticeException(Exception):
     """ An exception thrown for bad UpdateNotice data. """
     pass
@@ -445,7 +454,7 @@ class UpdateMetadata(object):
         else:   # obj is a file object
             infile = obj
 
-        for event, elem in iterparse(infile):
+        for event, elem in safe_iterparse(infile):
             if elem.tag == 'update':
                 try:
                     un = UpdateNotice(elem)
commit ee0fbe938cadea1457c89825412a978edf574125
Author: James Antill <james@and.org>
Date:   Thu Feb 10 15:54:59 2011 -0500

    Fix identifier typo.

diff --git a/yum/update_md.py b/yum/update_md.py
index 619250e..061a086 100644
--- a/yum/update_md.py
+++ b/yum/update_md.py
@@ -36,7 +36,7 @@ import rpmUtils.miscutils
 def safe_iterparse(filename):
     """ Works like iterparse, but hides XML errors (prints a warning). """
     try:
-        for event, elem in safe_iterparse(infile):
+        for event, elem in safe_iterparse(filename):
             yield event, elem
     except SyntaxError: # Bad XML
         print >> sys.stderr, "File is not valid XML:", filename
commit 82bb36e237e104454ccb6c9a5c412e983278f28c
Author: James Antill <james@and.org>
Date:   Thu Feb 10 15:55:37 2011 -0500

    Fix recursion problem, with safe_iterparse.

diff --git a/yum/update_md.py b/yum/update_md.py
index 061a086..39fa72e 100644
--- a/yum/update_md.py
+++ b/yum/update_md.py
@@ -36,7 +36,7 @@ import rpmUtils.miscutils
 def safe_iterparse(filename):
     """ Works like iterparse, but hides XML errors (prints a warning). """
     try:
-        for event, elem in safe_iterparse(filename):
+        for event, elem in iterparse(filename):
             yield event, elem
     except SyntaxError: # Bad XML
         print >> sys.stderr, "File is not valid XML:", filename
