commit a828573a879d1e3336b19b7a491fcdc43392f412
Author: James Antill <james@and.org>
Date:   Fri Feb 18 12:28:58 2011 -0500

    Add workaround for rhnplugin repos. setup, BZ 678043.

diff --git a/yum/__init__.py b/yum/__init__.py
index a3a1b02..4d185ff 100644
--- a/yum/__init__.py
+++ b/yum/__init__.py
@@ -574,6 +574,11 @@ class YumBase(depsolve.Depsolve):
 
             self.getReposFromConfig()
 
+        #  For rhnplugin, and in theory other stuff, calling
+        # .getReposFromConfig() recurses back into this function but only once.
+        # This means that we have two points on the stack leaving the above call
+        # but only one of them can do the repos setup. BZ 678043.
+        if hasattr(self, 'prerepoconf'):
             # Recursion
             prerepoconf = self.prerepoconf
             del self.prerepoconf
