diff --git a/yum/yumRepo.py b/yum/yumRepo.py
index 6ba058c..69e5ba4 100644
--- a/yum/yumRepo.py
+++ b/yum/yumRepo.py
@@ -796,15 +796,18 @@ class YumRepository(Repository, config.RepoConf):
                 verbose_logger.log(logginglevels.DEBUG_2, "Error getting package from media; falling back to url %s" %(e,))
 
         if url and scheme != "media":
-            ugopts = self._default_grabopts(cache=cache)
-            ug = URLGrabber(progress_obj = self.callback,
-                            copy_local = copy_local,
-                            reget = reget,
-                            failure_callback = self.failure_obj,
-                            interrupt_callback=self.interrupt_callback,
-                            checkfunc=checkfunc,
-                            size=size,
-                            **ugopts)
+#            ugopts = self._default_grabopts(cache=cache)
+#            ug = URLGrabber(progress_obj = self.callback,
+#                            copy_local = copy_local,
+#                            reget = reget,
+#                            failure_callback = self.failure_obj,
+#                            interrupt_callback=self.interrupt_callback,
+#                            checkfunc=checkfunc,
+#                            size=size,
+#                            **ugopts)
+#
+            ug = self._getgrab()
+            ug.opts.user_agent = self._default_grabopts(cache=cache)
 
             remote = url + '/' + relative
 
