diff --git a/yumRepo.py b/yumRepo.py
index 639a49a..8f16d7a 100644
--- a/yum/yumRepo.py
+++ b/yum/yumRepo.py
@@ -782,24 +782,9 @@ class YumRepository(Repository, config.RepoConf):
                 verbose_logger.log(logginglevels.DEBUG_2, "Error getting package from media; falling back to url %s" %(e,))

         if url and scheme != "media":
-            ug = URLGrabber(keepalive = self.keepalive,
-                            bandwidth = self.bandwidth,
-                            retry = self.retries,
-                            throttle = self.throttle,
-                            progress_obj = self.callback,
-                            copy_local = copy_local,
-                            reget = reget,
-                            proxies = self.proxy_dict,
-                            failure_callback = self.failure_obj,
-                            interrupt_callback=self.interrupt_callback,
-                            timeout=self.timeout,
-                            checkfunc=checkfunc,
-                            http_headers=headers,
-                            ssl_ca_cert = self.sslcacert
-                            )
-
+            ug = self._getgrab()
             ug.opts.user_agent = default_grabber.opts.user_agent
-
+
             remote = url + '/' + relative

             try:
